### PR TITLE
check pitch from application for user ptr

### PR DESCRIPTION
**Pull request by XinfengZhang**
Signed-off-by: XinfengZhang <carl.zhang@intel.com>

### CODE REVIEW

File: media_driver/linux/common/ddi/media_libva_util.cpp [hunk 1]
```diff
@@ -405,6 +411,23 @@ VAStatus DdiMediaUtil_AllocateSurface(
...
+    {
+        uint32_t gmmHAlign;
+        gmmHAlign = gmmResourceInfo->GetHAlign();
+        if((gmmPitch < mediaSurface->pSurfDesc->uiPitches[0]) && (((mediaSurface->pSurfDesc->uiPitches[0]-gmmPitch)%gmmHAlign) == 0))
+        {
+            gmmResourceInfo->OverridePitch(mediaSurface->pSurfDesc->uiPitches[0]);
+        }
+        else
+        {
+           DDI_ASSERTMESSAGE("the pitch does not match the basic requirement.");
```
Comment (dvrogozh) [inline media_driver/linux/common/ddi/media_libva_util.cpp]: This won't work for the case when user allocated pitch _matches_ the one from gmmlib:):). You need either use `<=` above and always call OverridePitch even if they match or explicitly avoid error with else `(gmmPitch != mediaSurface->pSurfDesc->uiPitches[0])`

File: media_driver/linux/common/ddi/media_libva_util.cpp [hunk 2]
```diff
@@ -405,6 +411,23 @@ VAStatus DdiMediaUtil_AllocateSurface(
...
     uint32_t    gmmHeight;
+
+    gmmPitch    = (uint32_t)gmmResourceInfo->GetRenderPitch();
+    if( DdiMediaUtil_IsExternalSurface(mediaSurface) && ( mediaSurface->pSurfDesc->uiVaMemType == VA_SURFACE_ATTRIB_MEM_TYPE_USER_PTR ) )
+    {
+        uint32_t gmmHAlign;
+        gmmHAlign = gmmResourceInfo->GetHAlign();
+        if((gmmPitch < mediaSurface->pSurfDesc->uiPitches[0]) && (((mediaSurface->pSurfDesc->uiPitches[0]-gmmPitch)%gmmHAlign) == 0))
+        {
+            gmmResourceInfo->OverridePitch(mediaSurface->pSurfDesc->uiPitches[0]);
```
Comment (dvrogozh) [inline media_driver/linux/common/ddi/media_libva_util.cpp]: I still don't understand why we need to call OverridePitch? If we notified gmmlib that memory is user allocated, from my perspective gmmlib should return user allocated pitch right away... So, I am not sure that this entire code block is actually needed.

File: media_driver/linux/common/ddi/media_libva_util.cpp [hunk 3]
```diff
@@ -405,6 +411,23 @@ VAStatus DdiMediaUtil_AllocateSurface(
     uint32_t    gmmPitch;
     uint32_t    gmmSize;
     uint32_t    gmmHeight;
+
+    gmmPitch    = (uint32_t)gmmResourceInfo->GetRenderPitch();
+    if( DdiMediaUtil_IsExternalSurface(mediaSurface) && ( mediaSurface->pSurfDesc->uiVaMemType == VA_SURFACE_ATTRIB_MEM_TYPE_USER_PTR ) )
+    {
+        uint32_t gmmHAlign;
+        gmmHAlign = gmmResourceInfo->GetHAlign();
```
Comment (dvrogozh) [inline media_driver/linux/common/ddi/media_libva_util.cpp]: I think this better be on the single line as:
int32_t gmmHAlign = gmmResourceInfo->GetHAlign();

File: media_driver/linux/common/ddi/media_libva_util.cpp [hunk 4]
```diff
@@ -354,6 +354,12 @@ VAStatus DdiMediaUtil_AllocateSurface(
                 gmmParams.BaseHeight = mediaSurface->pSurfDesc->uiSize / mediaSurface->pSurfDesc->uiPitches[0];
             }
         }
+        if(mediaSurface->pSurfDesc->uiVaMemType == VA_SURFACE_ATTRIB_MEM_TYPE_USER_PTR)
+        {
+            gmmParams.ExistingSysMemSize = mediaSurface->pSurfDesc->uiBuffserSize;
+            gmmParams.pExistingSysMem    = mediaSurface->pSurfDesc->ulBuffer;
+            gmmParams.Flags.Info.ExistingSysMem = true;
```
Comment (dvrogozh) [inline media_driver/linux/common/ddi/media_libva_util.cpp]: Surprisingly this regress some use cases for me (where I use LibVA VP to copy surface from userptr allocated to driver allocated). Specifically YUY2 and UYVY regressed. And some cases which I anticipated will be fixed (some of i420, yv12) were not fixed.

YUY2 regressed in the following way. Input surface is 176x144 and I allocated it as follows:
1. 16-alignment for width and 64-bytes alignment for pitch. I.e. width aligned to 192, stride=384
2. height was 4-aligned and remained 144

With the above allocation YUY2 worked just fine before the change. After the change we fail on a call to CreateResInfoObject(). Basically gmmlib for some reason wants height to be 154 while it is eventually not. I am trying to clarify with gmmlib folks where this requirement comes from.

I420 case actually follows the same YUY2 failure pattern. Before the change it just was copied incorrectly.

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 2 above)
Comment (XinfengZhang) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: 1. there are no interface to set the pitch from application except the "OverridePitch"

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 1 above)
Comment (XinfengZhang) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: yes,

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 2 above)
Comment (dvrogozh) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: In this case, how CreateResInfoObject() manages to validate anything? I can't imagine memory allocation validation knowing nothing about the pitch.

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 3 above)
Comment (XinfengZhang) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: sure

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 2 above)
Comment (XinfengZhang) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: yes, let me add engineer from https://github.com/intel/gmmlib , @drprajap, @kmvadsar

File: media_driver/linux/common/ddi/media_libva_util.cpp (hunk 4 above)
Comment (XinfengZhang) [reply media_driver/linux/common/ddi/media_libva_util.cpp]: @dvrogozh have you synced with gmmlib folks about it? why the alignment required 154 instead 144?

**Pull request merged**